File: crates/goose-sdk/examples/uniffi/PingAaif.kt
package examples

import uniffi.goose_sdk.Agent
import uniffi.goose_sdk.EventSink
import uniffi.goose_sdk_types.AgentEvent
import uniffi.goose_sdk_types.ExtensionSpec
import uniffi.goose_sdk_types.ProviderSpec

private object Style {
    const val DIM = "\u001B[2m"
    const val CYAN = "\u001B[36m"
    const val GREEN = "\u001B[32m"
    const val RED = "\u001B[31m"
    const val RESET = "\u001B[0m"
}

private fun String.paint(color: String) = "$color$this${Style.RESET}"

private fun String.preview(maxLines: Int = 3, maxWidth: Int = 100): String =
    lineSequence()
        .filter { it.isNotBlank() }
        .map { it.take(maxWidth) }
        .take(maxLines)
        .joinToString("\n  ")

private class Printer : EventSink {
    private var midText = false

    override fun onEvent(event: AgentEvent) {
        when (event) {
            is AgentEvent.AssistantText -> {
                print(event.text)
                midText = true
            }
            is AgentEvent.ToolRequest -> {
                endTextLine()
                val args = event.arguments.replace("\n", " ").take(120)
                println("→ ${event.name}".paint(Style.CYAN) + " " + args.paint(Style.DIM))
            }
            is AgentEvent.ToolResponse -> {
                endTextLine()
                val color = if (event.isError) Style.RED else Style.GREEN
                val marker = if (event.isError) "✗" else "✓"
                println(marker.paint(color) + " " + event.output.preview().paint(Style.DIM))
                println()
            }
            is AgentEvent.Thinking -> Unit
        }
        System.out.flush()
    }

    override fun onError(error: String) {
        System.err.println("\n${"error:".paint(Style.RED)} $error")
    }

    override fun onDone() = endTextLine()

    private fun endTextLine() {
        if (midText) {
            println()
            midText = false
        }
    }
}

fun main() {
    System.err.println("configuring agent…".paint(Style.DIM))

    val agent = Agent().apply {
        configure(
            ProviderSpec(
                name = System.getenv("GOOSE_PROVIDER"),
                model = System.getenv("GOOSE_MODEL"),
            ),
            listOf(ExtensionSpec.Builtin(name = "developer")),
        )
    }

    System.err.println("> ping aaif.io".paint(Style.DIM) + "\n")
    agent.reply("ping aaif.io", Printer())
}

File: crates/goose-sdk/examples/uniffi/ping_aaif.py
"""Minimal goose SDK demo: ask the agent to ping aaif.io."""

from __future__ import annotations

import os
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
sys.path.insert(0, str(HERE.parent.parent))

from generated.goose_sdk import Agent, EventSink  # noqa: E402
from generated.goose_sdk_types import AgentEvent, ExtensionSpec, ProviderSpec  # noqa: E402

DIM = "\033[2m"
CYAN = "\033[36m"
GREEN = "\033[32m"
RED = "\033[31m"
RESET = "\033[0m"


def paint(color: str, text: str) -> str:
    return f"{color}{text}{RESET}"


def preview(output: str, max_lines: int = 3, max_width: int = 100) -> str:
    lines = (line[:max_width] for line in output.splitlines() if line.strip())
    return "\n  ".join(list(lines)[:max_lines])


class Printer(EventSink):
    def __init__(self) -> None:
        self._mid_text = False

    def on_event(self, event: AgentEvent) -> None:
        if isinstance(event, AgentEvent.ASSISTANT_TEXT):
            print(event.text, end="", flush=True)
            self._mid_text = True
            return

        self._end_text_line()

        if isinstance(event, AgentEvent.TOOL_REQUEST):
            args = event.arguments.replace("\n", " ")[:120]
            print(f"{paint(CYAN, '→ ' + event.name)} {paint(DIM, args)}", flush=True)

        elif isinstance(event, AgentEvent.TOOL_RESPONSE):
            color = RED if event.is_error else GREEN
            marker = "✗" if event.is_error else "✓"
            print(f"{paint(color, marker)} {paint(DIM, preview(event.output))}\n", flush=True)

    def on_error(self, error: str) -> None:
        print(f"\n{paint(RED, 'error:')} {error}", file=sys.stderr)

    def on_done(self) -> None:
        self._end_text_line()

    def _end_text_line(self) -> None:
        if self._mid_text:
            print()
            self._mid_text = False


def main() -> None:
    print(paint(DIM, "configuring agent…"), file=sys.stderr)

    agent = Agent()
    agent.configure(
        ProviderSpec(
            name=os.environ.get("GOOSE_PROVIDER"),
            model=os.environ.get("GOOSE_MODEL"),
        ),
        [ExtensionSpec.BUILTIN(name="developer")],
    )

    print(paint(DIM, "> ping aaif.io") + "\n", file=sys.stderr)
    agent.reply("ping aaif.io", Printer())


if __name__ == "__main__":
    main()
